### PR TITLE
[InstCombine] Fold (X / C) < X and (X >> C) < X into X > 0

### LINKED ISSUE #85313: missed fold, x / c < x => x > 0
https://alive2.llvm.org/ce/z/yG_wxp
https://godbolt.org/z/fKo893Pv1

```llvm
define i1 @src(i32 %0, i1 %1, i1 %2, i1 %3, i8 %4) {
if.then160:
  %5 = sdiv i32 %0, 131313
  %6 = icmp slt i32 %5, %0
  ret i1 %6
}


define i1 @tgt(i32 %0, i1 %1, i1 %2, i1 %3, i8 %4) {
if.then160:
  %5 = icmp sgt i32 %0, 0
  ret i1 %5
}
```

Constant 131313 can be anything but 0 or 1.

@dtcxzyw @regehr

### PR DESCRIPTION
**Pull request by Poseydon42**
Proofs:
https://alive2.llvm.org/ce/z/52droC
This resolves #85313

### CODE REVIEW

**Comment by goldsteinn** [comment]
Just a note on the proofs.
You really only proved for the exact constant chosen i.e `lshr x, 1` in the case of the `lshr` proofs. That doesn't necessarily transfer to all other constants (in fact 0 is incorrect).
See some general proofs: https://alive2.llvm.org/ce/z/3YWXgD
I filled in the `sdiv` cases, can you please complete the `udiv`/`lshr`/`ashr` ones.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7417]: Divisor->ugt(1)

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7418]: You are missing a check that in the incoming Predicate was `ICMP_SLT`.

In general though, can you handle a lot more predicates and the general pattern `ICmpInst::getSwappedPredicate(I.getPredicate())`

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7418]: Also note you can handle just about any predicate here.
For `sdiv`/`ashr`, you need `!ICmpInst::isUnsigned(I.getPredicate())`. For `udiv`/`lshr`, any predicate works.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7428]: `M_Deferred` isn't really right here. Thats for if you are matching a specific `V` set earlier in the same `match` statement.
Here you want to use `m_Specific`.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7433]: This code should probably all be in the `foldICmpCommutative` function.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7416]: Really this can just `Op1 == Dividend` (likewise below in the `lshr` case).

**Comment by goldsteinn** [comment]
Also welcome and thank you for the commit. There a rough draft of a contribution guide for `InstCombine` here: #79007

You might find it useful.

**Comment by Poseydon42** [comment]
I have modified the code to work with all predicates for `udiv`/`lshr` and unsigned predicates for `sdiv`, however I believe that `ashr` does not work with this type of fold. As you can see [here](https://alive2.llvm.org/ce/z/cVMB9e), the only two predicates that it does work with are `slt` and `sge`. Should I include these two folds for `ashr` or should I ignore it altogether and only focus on the other 3 instructions?

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7417]: This is buggy I think. Imagine a non-splat vec with denom `<i8 0, i8 123>`. Not zero/ones but still incorrect. Think you should change `m_ImmConstant` to `m_APInt` then just use `ugt(1)`

**Comment by goldsteinn** [comment]
> I have modified the code to work with all predicates for `udiv`/`lshr` and unsigned predicates for `sdiv`, however I believe that `ashr` does not work with this type of fold. As you can see [here](https://alive2.llvm.org/ce/z/cVMB9e), the only two predicates that it does work with are `slt` and `sge`. Should I include these two folds for `ashr` or should I ignore it altogether and only focus on the other 3 instructions?

Think made a mistake when merging your updates.

**Comment by Poseydon42** [comment]
Yeah, that merge was an accident, I'll remove it before pushing the most recent changes. It seems to me like handling `ashr` would just unnecessarily increase the complexity of the code without providing significant benefit, so I'll omit it. The only question I have now is whether I should add tests for all three instructions I'm working with and all 10 (or 6) comparison predicates, making it 26 tests in total, which, together with negative tests, will bring it closer to 30, or should I simply test 2-3 predicates for each of the three instructions?

**Comment by goldsteinn** [comment]
> Yeah, that merge was an accident, I'll remove it before pushing the most recent changes. It seems to me like handling `ashr` would just unnecessarily increase the complexity of the code without providing significant benefit, so I'll omit it. The only question I have now is whether I should add tests for all three instructions I'm working with and all 10 (or 6) comparison predicates, making it 26 tests in total, which, together with negative tests, will bring it closer to 30, or should I simply test 2-3 predicates for each of the three instructions?

So generally you should be testing different control flows moreso than code.

In this case you have all predicates have the same control flow for `lshr`/`udiv`, so maybe between the two test one equality pred, one signed, and one unsigned (i.e eq, sle, ugt). Then a negative test for each maybe with an out of bounds constant.

For `sdiv` maybe ne / sgt with two negative tests  (out of bounds/unsigned pred).

Then you can pretty randomly make 1-2 of the tests use vec.

Please be sure to have a negative test with a vec where one vec element is invalid and the other is valid.

**Comment by Poseydon42** [comment]
Updated proofs (26 in total) can be found [here](https://alive2.llvm.org/ce/z/TDRxuG). I've fixed most of the issues mentioned here, the only two things I'm worried about are:
  1.The current code does not work with vector constants that have different values in each component (e.g. take a look at test case `lshr_by_const_cmp_sle_value`). I believe this is caused by `m_APInt` not matching when a vector constant has different components, but I am unsure about that. Can something be done about this or should I leave this particular scenario unaffected by the optimization?
  2. It appears that `lshr` already has a similar fold implemented, since test case `lshr_by_const_cmp_ule_value` was not affected by the changes that I've made (it folded to `ret i1 true` both before and after the changes to the code). Can someone confirm that this is the case, and if so, should I remove the `lshr` code path from my PR?

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:14]: This shouldn't be needed.

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:14]: Likewise below.

**Comment by goldsteinn** [comment]
> Updated proofs (26 in total) can be found [here](https://alive2.llvm.org/ce/z/TDRxuG). I've fixed most of the issues mentioned here, the only two things I'm worried about are: 1.The current code does not work with vector constants that have different values in each component (e.g. take a look at test case `lshr_by_const_cmp_sle_value`). I believe this is caused by `m_APInt` not matching when a vector constant has different components, but I am unsure about that. Can something be done about this or should I leave this particular scenario unaffected by the optimization? 

Yes that is one of the drawbacks of APInt, it only match if the vector is a splat (all elements that same).
You could use: `m_SpecificInt_ICMP` (see PatternMatch.h).
So for the div cases: `m_SpecificInt_ICMP(ICmpInst::ICMP_UGT, APInt::getOneBitSet(Op0->getType()->getScalarSizeInBits(), 1))` (not the cleanest API to use).

> 2. It appears that `lshr` already has a similar fold implemented, since test case `lshr_by_const_cmp_ule_value` was not affected by the changes that I've made (it folded to `ret i1 true` both before and after the changes to the code). Can someone confirm that this is the case, and if so, should I remove the `lshr` code path from my PR?

See: https://github.com/llvm/llvm-project/blob/main/llvm/lib/Analysis/InstructionSimplify.cpp#L3172
No need to go out of your way to avoid that case. Its not extra code on your end to handle (and its correct). I wouldn't use it as a test, however, as its not going through your code path.

**Comment by Poseydon42** [comment]
Right, after a painful hour of rewriting git history multiple times, it seems to me that I've covered everything that was mentioned here before. Please let me know if there's anything else for me to fix or improve.

**Comment by Poseydon42** [comment]
Oops, looks like I forgot to run tests on my machine before pushing the changes. I've localised the source of one of many failing tests to this function in `getelementptr.ll`:
```
define i1 @test10(ptr %x, ptr %y) {
; CHECK-LABEL: @test10(
; CHECK-NEXT:    [[T4:%.*]] = icmp eq ptr [[X:%.*]], [[Y:%.*]]
; CHECK-NEXT:    ret i1 [[T4]]
;
  %t1 = getelementptr { i32, i32 }, ptr %x, i32 0, i32 1
  %t3 = getelementptr { i32, i32 }, ptr %y, i32 0, i32 1
  %t4 = icmp eq ptr %t1, %t3
  ret i1 %t4
}
```

The crash report indicates assertion failure in `APInt.h` within `setBit()`. It appears to be caused by the fact that for `ptr` types, bit size is reported as 0, so later in the matching code, a call to `APInt::getOneBitSet` essentially asks for a number that has 0 bits, with bit #0 set. I'm unsure if I should simply ignore cases like this (e.g. by adding `if (Op0->getType()->getScalarSizeInBits() == 0) return nullptr`), or if there's a nicer/more usual way of handling this?

**Comment by goldsteinn** [comment]
> Oops, looks like I forgot to run tests on my machine before pushing the changes. I've localised the source of one of many failing tests to this function in `getelementptr.ll`:
> 
> ```
> define i1 @test10(ptr %x, ptr %y) {
> ; CHECK-LABEL: @test10(
> ; CHECK-NEXT:    [[T4:%.*]] = icmp eq ptr [[X:%.*]], [[Y:%.*]]
> ; CHECK-NEXT:    ret i1 [[T4]]
> ;
>   %t1 = getelementptr { i32, i32 }, ptr %x, i32 0, i32 1
>   %t3 = getelementptr { i32, i32 }, ptr %y, i32 0, i32 1
>   %t4 = icmp eq ptr %t1, %t3
>   ret i1 %t4
> }
> ```
> 
> The crash report indicates assertion failure in `APInt.h` within `setBit()`. It appears to be caused by the fact that for `ptr` types, bit size is reported as 0, so later in the matching code, a call to `APInt::getOneBitSet` essentially asks for a number that has 0 bits, with bit #0 set. I'm unsure if I should simply ignore cases like this (e.g. by adding `if (Op0->getType()->getScalarSizeInBits() == 0) return nullptr`), or if there's a nicer/more usual way of handling this?

I'd say just handle the splat vector case for now. If #85676 gets in before this, you could use that API.

**Comment by Poseydon42** [comment]
> I'd say just handle the splat vector case for now. If #85676 gets in before this, you could use that API.
Done, should be passing tests now.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7111]:
```suggestion
    if (match(Op0, m_UDiv(m_Specific(Op1), m_APInt(Divisor))) &&  Divisor->ugt(1)) {
```

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7119]: This code seems unreachable :(

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7129]:
```suggestion
    if (match(Op0, m_LShr(m_Specific(Op1), m_APInt(Shift))) &&
        !Shift->isZero()) {
```

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7119]: Do you mean `m_SDiv`?

File: llvm/test/Transforms/InstCombine/icmp-div-constant.ll
Comment (goldsteinn) [inline llvm/test/Transforms/InstCombine/icmp-div-constant.ll:462]: to make any of these tests meaningful, they really need to be in the form of a non-splat vector constant where one element is the identity, and another is some interesting constant.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7124]: nit: in your comments 'cmp' -> 'pred'.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7134]: Does it hold for `ashr`?
An example: https://alive2.llvm.org/ce/z/peP7WS

File: llvm/test/Transforms/InstCombine/icmp-div-constant.ll
Comment (Poseydon42) [reply llvm/test/Transforms/InstCombine/icmp-div-constant.ll:462]: The problem is, non-splat vectors won't be touched by my code either as they will fail the match due to the issue described above.

**Review by dtcxzyw — CHANGES_REQUESTED** [review]
https://github.com/llvm/llvm-project/pull/79007/files may help :)

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (Poseydon42) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7134]: Only with `slt` and `sge`, so I thought it not being worth implementing, as the potential gain seems minimal.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (Poseydon42) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7119]: Yes, looks like I copy-pasted the `match` arguments from the code above and forgot to change it to `sdiv`.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7134]: It makes sense to me.

**Comment by Poseydon42** [comment]
Updated proofs with `ashr`: https://alive2.llvm.org/ce/z/52droC

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7130]: nit: `!Shift->isZero()`

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7135]: nit: There is a bit of redundancy in that each of the return cases is identical, but at least imo its fine.

**Comment by goldsteinn** [comment]
LGTM, wait on one more approval.

**Comment by Poseydon42** [comment]
Ping @nikic @dtcxzyw

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7116]:
```suggestion
    if (!ICmpInst::isUnsigned(Pred) && match(Op0, m_SDiv(m_Specific(Op1), m_APInt(Divisor))) &&
        Divisor->ugt(1)) {
```
`!ICmpInst::isUnsigned(Pred)` is cheaper.

Comment (Poseydon42) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp]: Fixed.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp]: It is not fixed.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7106]: nit: `~pred` looks like the inverse predicate, not the swapped predicate, which are different things. I'd write this either as
```suggestion
  // Folding (X / Y) pred X => 0 pred X for some constant Y other than 0 or 1
```
or
```suggestion
  // Folding (X / Y) pred X => X swap(pred) 0 for some constant Y other than 0 or 1
```

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7110]: Worth noting that this condition could be != 1 rather than > 1, because division by zero is UB anyway. Ultimately makes no difference in terms of behavior though, so your variant is fine as well.

File: llvm/test/Transforms/InstCombine/icmp-div-constant.ll
Comment (nikic) [inline llvm/test/Transforms/InstCombine/icmp-div-constant.ll:424]:
```suggestion
define i1 @lshr_by_const_cmp_sge_value(i32 %x) {
```

**Review by nikic — APPROVED** [review]
LGTM

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (Poseydon42) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp]: I believe it is fixed now.

**Comment by Poseydon42** [comment]
Thanks nikic, looks like it is now ready for another review.
Ping @dtcxzyw

**Pull request merged**
